### PR TITLE
support for snapshot 24w04a (1.20.5)

### PR DESCRIPTION
**Pull request by Outfluencer**
related to #3601
if you spot bugs please report

### CODE REVIEW

**Comment by Outfluencer** [comment]
thoghts: maybe add transfer api and cookie api, also maybe provide config options to decline transfers

**Comment by Outfluencer** [comment]
Notice: if you want to test with an vanilla server (i think there are now spigot snapshot build) you need to change the uniqueId in 
handle(final EncryptionResponse encryptResponse)

to the offline uuid 

`if ( obj != null && obj.getId() != null )
{
    loginProfile = obj;
    name = obj.getName();
    uniqueId = Util.getUUID( obj.getId() );
    if ( VANILLA_TESTING )
    {
        // otherwise we get kicked for invalid signature on vanilla
        uniqueId = UUID.nameUUIDFromBytes( ( "OfflinePlayer:" + getLoginRequest().getData() ).getBytes() );
    }
    finish();
    return;
}`

**Comment by Outfluencer** [comment]
@md-5 i thought about the  transfer api, should we store a cookie inside the client like "bungeecord:transfer" with a json as data like the BungeeCord version and build number and a timestamp, before we sent the transfer packet to the player?

Whats your opinion on this one?

**Comment by OpticFusion1** [comment]
> @md-5 i thought about the transfer api, should we store a cookie inside the client like "bungeecord:transfer" with a json as data like the BungeeCord version and build number and a timestamp, before we sent the transfer packet to the player?
> 
> Whats your opinion on this one?

I personally see no reason for this.

**Comment by NEZNAMY** [comment]
Using `sendPacket` may cause the plugin to be sent during wrong phase (such as configuration).

**Comment by Outfluencer** [comment]
> Using `sendPacket` may cause the packet to be sent during wrong phase (such as configuration).

what exactly do you mean?

**Comment by Outfluencer** [comment]
the cookie packets are present in game login and config state and the transfer packet witch api is only after login state is registered for configuration and game, its not possible that the packet is in the wronge phase

**Comment by md-5** [comment]
> Notice: if you want to test with an vanilla server (i think there are no spigot snapshot build) you need to change the uniqueId in handle(final EncryptionResponse encryptResponse) to the offline uuid
> 
> ```
> Callback<String> handler = new Callback<String>()
> {
>     @Override
>     public void done(String result, Throwable error)
>     {
>         if ( error == null )
>         {
>             LoginResult obj = BungeeCord.getInstance().gson.fromJson( result, LoginResult.class );
>             if ( obj != null && obj.getId() != null )
>             {
>                 loginProfile = obj;
>                 name = obj.getName();
>                 uniqueId = Util.getUUID( obj.getId() );
>                 // here is the change
>                 if ( VANILLA_TESTING )
>                 {
>                     // otherwise we get kicked for invalid signature on vanilla
>                     uniqueId = UUID.nameUUIDFromBytes( ( "OfflinePlayer:" + getLoginRequest().getData() ).getBytes() );
>                 }
>                 finish();
>                 return;
>             }
>             disconnect( bungee.getTranslation( "offline_mode_player" ) );
>         } else
>         {
>             disconnect( bungee.getTranslation( "mojang_fail" ) );
>             bungee.getLogger().log( Level.SEVERE, "Error authenticating " + getName() + " with minecraft.net", error );
>         }
>     }
> };
> ```
I guess this only applies if bungee is online mode? Should probably look into this and if it can be supported without an option as Bungee should support vanilla.

**Comment by md-5** [comment]
> @md-5 i thought about the transfer api, should we store a cookie inside the client like "bungeecord:transfer" with a json as data like the BungeeCord version and build number and a timestamp, before we sent the transfer packet to the player?
> 
> Whats your opinion on this one? Do you have any idea for Cookie standard format, that may also be used in spigot

I don't think that's necessary, but there should be an API

**Comment by Outfluencer** [comment]
> > Notice: if you want to test with an vanilla server (i think there are no spigot snapshot build) you need to change the uniqueId in handle(final EncryptionResponse encryptResponse) to the offline uuid
> 
> > 
> 
> > ```
> 
> > Callback<String> handler = new Callback<String>()
> 
> > {
> 
> >     @Override
> 
> >     public void done(String result, Throwable error)
> 
> >     {
> 
> >         if ( error == null )
> 
> >         {
> 
> >             LoginResult obj = BungeeCord.getInstance().gson.fromJson( result, LoginResult.class );
> 
> >             if ( obj != null && obj.getId() != null )
> 
> >             {
> 
> >                 loginProfile = obj;
> 
> >                 name = obj.getName();
> 
> >                 uniqueId = Util.getUUID( obj.getId() );
> 
> >                 // here is the change
> 
> >                 if ( VANILLA_TESTING )
> 
> >                 {
> 
> >                     // otherwise we get kicked for invalid signature on vanilla
> 
> >                     uniqueId = UUID.nameUUIDFromBytes( ( "OfflinePlayer:" + getLoginRequest().getData() ).getBytes() );
> 
> >                 }
> 
> >                 finish();
> 
> >                 return;
> 
> >             }
> 
> >             disconnect( bungee.getTranslation( "offline_mode_player" ) );
> 
> >         } else
> 
> >         {
> 
> >             disconnect( bungee.getTranslation( "mojang_fail" ) );
> 
> >             bungee.getLogger().log( Level.SEVERE, "Error authenticating " + getName() + " with minecraft.net", error );
> 
> >         }
> 
> >     }
> 
> > };
> 
> > ```
> 
> I guess this only applies if bungee is online mode? Should probably look into this and if it can be supported without an option as Bungee should support vanilla.

Yes only for online mode, for offline mode it works fine as the uuids are the same, in online mode i think some signatures are broken in some packets because of the uuid difference

**Comment by Janmm14** [comment]
an api would need to be added to an interface

**Comment by Outfluencer** [comment]
@md-5 you can achieve vanilla compatibility by dropping the Chat Session packet if enforce secure profile is set to false in bungee, if you do, you also can send messages to the server, but when the message is sent from the server to all clients they will show this: 
image

**Comment by md-5** [comment]
> @md-5 you can achieve vanilla compatibility by dropping the Chat Session packet if enforce secure profile is set to false in bungee, if you do, you also can send messages to the server, but when the message is sent from the server to all clients they will show this: image

If we do the fix you initially mentioned does the chat error show?

**Comment by Outfluencer** [comment]
> > @md-5 you can achieve vanilla compatibility by dropping the Chat Session packet if enforce secure profile is set to false in bungee, if you do, you also can send messages to the server, but when the message is sent from the server to all clients they will show this: image
> 
> If we do the fix you initially mentioned does the chat error show?

i think it does not show

**Comment by md-5** [comment]
> > > @md-5 you can achieve vanilla compatibility by dropping the Chat Session packet if enforce secure profile is set to false in bungee, if you do, you also can send messages to the server, but when the message is sent from the server to all clients they will show this: image
> > 
> > 
> > If we do the fix you initially mentioned does the chat error show?
> 
> as i remember it does not show

Can you please check? And then perhaps make a separate PR but replace `VANILLA_TESTING` with `(onlineMode && !ipForward)`?

**Comment by Outfluencer** [comment]
Okay i can do it when i am home in like 2 hours

**Comment by Outfluencer** [comment]
nah i forget i got my laptop on my, i tested, its working

**Comment by md-5** [comment]
Based on an initial review, looks good to me. Well done!

**Comment by Outfluencer** [comment]
What do you guys think about adding an Player Transfer Event, so we can cancel transfers by the backend server?

File: api/src/main/java/net/md_5/bungee/api/connection/PendingConnection.java [hunk 1]
```diff
@@ -89,4 +90,15 @@ public interface PendingConnection extends Connection
...
+    /**
+     * Retrieves a cookie from this pending connection.
+     *
+     * @param cookie the resource location of the cookie, for example "bungeecord:my_cookie"
+     *
+     * @return a {@link CompletableFuture} that will be completed when the Cookie response is received
+     * Only useable for 1.20.5 clients or newer during the login, config or game phase
+     * if the cookie is not set in the client, the {@link CompletableFuture} will complete with a null value
+     */
+    CompletableFuture<byte[]> retrieveCookie(String cookie);
```
Comment (md-5) [inline api/src/main/java/net/md_5/bungee/api/connection/PendingConnection.java:103]: Nice use of CompletableFuture!

Can you please add `@ApiStatus.Experimental` annotation?

You'll need to merge in master as I just added the dependency in c69acf728c75b98070c225595bccc682ded13b74

File: api/src/main/java/net/md_5/bungee/api/connection/ProxiedPlayer.java [hunk 2]
```diff
@@ -339,4 +340,36 @@ public enum MainHand
...
+     * Retrieves a cookie from this player.
+     *
+     * @param cookie the resource location of the cookie, for example "bungeecord:my_cookie"
+     *
+     * @return a {@link CompletableFuture} that will be completed when the Cookie response is received
+     * Only useable for 1.20.5 clients or newer,
+     * if the cookie is not set in the client, the {@link CompletableFuture} will complete with a null value
+     */
+    CompletableFuture<byte[]> retrieveCookie(String cookie);
+
```
Comment (md-5) [inline api/src/main/java/net/md_5/bungee/api/connection/ProxiedPlayer.java:354]: Remove extra newlines and add experimental annotations as discussed above to all these new API methods

File: api/src/main/java/net/md_5/bungee/api/connection/ProxiedPlayer.java [hunk 3]
```diff
@@ -339,4 +340,36 @@ public enum MainHand
...
+     *
+     * @param cookie the resource location of the cookie, for example "bungeecord:my_cookie"
+     * @param data the data to store in the cookie
+     *
+     * Only useable for 1.20.5 clients or newer
+     */
+    void storeCookie(String cookie, byte[] data);
+
+    /**
+     * Transfers this player to a different server specified by host and port.
```
Comment (md-5) [inline api/src/main/java/net/md_5/bungee/api/connection/ProxiedPlayer.java:367]: Please be more specific that it asks the client to transfer, it's not a bungee transfer

File: protocol/src/main/java/net/md_5/bungee/protocol/packet/CookieResponse.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+{
+
+    private String cookie;
+    private byte[] data;
+
+    @Override
+    public void read(ByteBuf buf, ProtocolConstants.Direction direction, int protocolVersion)
+    {
+        cookie = readString( buf );
+        if ( buf.readBoolean() )
```
Comment (md-5) [inline protocol/src/main/java/net/md_5/bungee/protocol/packet/CookieResponse.java:26]: I think you can use the methods `readNullable` and `writeNullable` with `ByteBuf::readBoolean` etc

File: protocol/src/main/java/net/md_5/bungee/protocol/packet/Login.java [hunk 5]
```diff
@@ -135,7 +136,13 @@ public void read(ByteBuf buf, ProtocolConstants.Direction direction, int protoco
             dimension = readString( buf );
             worldName = readString( buf );
             seed = buf.readLong();
-            gameMode = buf.readUnsignedByte();
+            if ( protocolVersion >= ProtocolConstants.MINECRAFT_1_20_5 )
```
Comment (md-5) [inline protocol/src/main/java/net/md_5/bungee/protocol/packet/Login.java:139]: Can probably just leave this, don't think it's actually changed and will never matter

File: proxy/src/main/java/net/md_5/bungee/conf/Configuration.java [hunk 6]
```diff
@@ -68,6 +68,8 @@ public class Configuration implements ProxyConfig
     private int compressionThreshold = 256;
     private boolean preventProxyConnections;
     private boolean forgeSupport;
+    private boolean rejectTransfers;
+
```
Comment (md-5) [inline proxy/src/main/java/net/md_5/bungee/conf/Configuration.java:72]: Extra newline

File: proxy/src/main/java/net/md_5/bungee/UserConnection.java [hunk 7]
```diff
@@ -771,4 +774,24 @@ public Scoreboard getScoreboard()
...
+    @Override
+    public CompletableFuture<byte[]> retrieveCookie(String cookie)
+    {
+        return pendingConnection.retrieveCookie( cookie );
+    }
+
+    @Override
+    public void storeCookie(String cookie, byte[] data)
+    {
+        Preconditions.checkState( getPendingConnection().getVersion() >= ProtocolConstants.MINECRAFT_1_20_5, "Cookies are only supported in 1.20.5 and above" );
```
Comment (md-5) [inline proxy/src/main/java/net/md_5/bungee/UserConnection.java:787]: Something to discuss - whether we should force a namespace

**Review by md-5 — COMMENTED** [review]
Thanks, looks really good. Some minor comments.

File: proxy/src/main/java/net/md_5/bungee/UserConnection.java (hunk 7 above)
Comment (Outfluencer) [reply proxy/src/main/java/net/md_5/bungee/UserConnection.java:787]: for storeCookie, we don't need to as the client creates a "minecraft:" namespace if no namespace is sent, thats why i added "minecraft:" as default namespace for retrieveCookie so we can run code like this without any issues

```
player.storeCookie("test", "Outfluencer".getBytes());
player.retrieveCookie("test").thenAccept((data) -> {
    player.sendMessage(new String(data));
});
````

File: api/src/main/java/net/md_5/bungee/api/connection/PendingConnection.java (hunk 1 above)
Comment (Outfluencer) [reply api/src/main/java/net/md_5/bungee/api/connection/PendingConnection.java:104]: done

File: proxy/src/main/java/net/md_5/bungee/UserConnection.java (hunk 7 above)
Comment (md-5) [reply proxy/src/main/java/net/md_5/bungee/UserConnection.java:787]: Yep, I know. But maybe for good API we want to force a namespace

**Review by md-5 — APPROVED** [review]

**Comment by md-5** [comment]
Thanks!

FYI I made some minor changes to the cookie handling in InitialHandler to make it properly thread safe given it's a future and may commonly be used async. Please let me know if any concerns.

**Comment by md-5** [comment]
Should we add an isTransferred to PendingConnection ?

**Pull request closed without merging**